doc/csharp/csharp-null-ref-types-proposal.md

```diff
@@ -0,0 +1,197 @@
+ Handling of null reference types in C# code generation
+----
+* Author(s): Tony Newell
+* Approver: 
+* Status: Draft
+* Last updated: June 2023
+
+## Abstract
+
+The protocol buffers compiler (`protoc`) and the gRPC C# compiler plugin 
+(`grpc_csharp_plugin`) generate C# code that does not use the nullable reference
+ types language features introduced in C# 8.0.
+
+This proposal discusses why there are no plans to explicitly handle nullable 
+reference types in the generated code.
+
+## Background
+
+A note on the terminology used in this document for the various *.NET* versions:
+- *.NET Framework* - the implementation of *.NET* that runs only on Windows, such 
+  as *.NET Framework 4.8.1*
+- *.NET Core* - within this document, references to *.NET Core* are intended to include
+ *.NET Core 3.1*, *.NET 5*, and later. Strictly Microsoft refers to all versions
+ from *.NET 5* and later as just *.NET*. 
+
+*Nullable reference types* refers to a group of optional features introduced 
+in C# 8.0 to enable the compiler to do static flow analysis to determine if a
+variable might be null before it is dereferenced. They are compile time features
+that do not affect the runtime.
+
+Various issues have been raised asking for the code generated by the protocol
+buffers compiler (`protoc`) and the gRPC C# compiler plugin 
+(`grpc_csharp_plugin`) to support nullable reference types, including:
+
+- [Investigate support for nullable reference types in generated code](https://github.com/grpc/grpc-dotnet/issues/606)
+- [Add option to generate C# nullable annotations](https://github.com/protocolbuffers/protobuf/issues/6632)
+- [C#: Consider exposing protobuf optional fields as nullable properties](https://github.com/protocolbuffers/protobuf/issues/9083)
+- [Support C# 8 nullable type references in Grpc.Tools](https://github.com/grpc/grpc/issues/20729)
+
+## Proposal
+
+The proposal is to maintain the status quo and to keep code generation 
+unchanged. Reasons are given below.
+
+## Rationale
+
+The reasons for currently maintaining the status quo are summarised here:
+1. The generated code is not in a nullable context
+2. *.NET Framework* projects default to using C# 7.3
+3. Building gRPC projects with nullable enabled works today
+4. Breaking backwards compatibility with gRPC service classes
+5. Keeping the generated code simple and consistent
+
+There are very few disadvantages to not supporting nullable reference types in
+the generated code. These are given later.
+
+The above reasons are explained in more detail:
+
+### 1. The generated code is not in a nullable context
+
+The generated code is not in a nullable context. The `// <auto-generated>` 
+comment in the code disables it. No nullable checks are made and no compiler 
+warnings are generated for this code, even if the code is compiled in a project 
+with nullable checks enabled.
```

Comment:
Yes I know. The point was that the code generated today is compiler with nullable context disabled thus it compiles OK with no warnings, as in nothing needs to be "fixed" to get it compiling cleanly today.

Reply:
The goal is to give gRPC users information about null references in their code. For example:

```proto
message User {
  string name = 1;
  Address address = 2;
}

message Address {
  string street = 1;
  string city = 2;
}
```

```cs
public override Task<Empty> AddUser(User user, ServerCallContext context)
{
    AddressEntity entity = new
    {
        Street = user.Address.Street, // Possible null reference exception
        City = user.Address.City,
    };
    _addressRepository.Add(entity);

    // rest of the method
}
```

Right now this will happily compile without warnings. If gRPC code gen had annotations then they would get a warning and would know that `User.Address` could be null.

```cs
public override Task<Empty> AddUser(User user, ServerCallContext context)
{
    if (user.Address != null)
    {
        var entity = new AddressEntity
        {
            Street = user.Address.Street,
            City = user.Address.City,
        };
        _addressRepository.Add(entity);
    }

    // rest of the method
}
```